console is trying to run exe under Linux

```
I 260714 195139.355 [AWT-EventQueue-0] StatusWindow - Linux 5.15.0-181-generic
...
I 260714 195150.988 [class com.rusefi.maintenance.DfuFlasher thread] FindFileHelper - Scanning /home/work/rusefi/soft/rusefi.snapshot.uaefi_pro_h7/console/.. for rusefi_/.bin
I 260714 195150.989 [class com.rusefi.maintenance.DfuFlasher thread] FindFileHelper - Searching for rusefi_*.bin in /home/work/rusefi/soft/rusefi.snapshot.uaefi_pro_h7/console/..
I 260714 195150.989 [class com.rusefi.maintenance.DfuFlasher thread] FindFileHelper - Located ../rusefi_development_yymmdd_uaefi_pro_h7_1976663299_local.bin
I 260714 195151.002 [class com.rusefi.maintenance.DfuFlasher thread] ExecHelper - executeCommand java.io.IOException: Cannot run program "./STM32_Programmer_CLI/bin/STM32_Programmer_CLI.exe" (in directory "./STM32_Programmer_CLI/bin"): Exec failed, error: 2 (No such file or directory) 
java.lang.RuntimeException: com.rusefi.maintenance.ErrorExecutingCommand
	at com.rusefi.maintenance.ExecHelper.executeCommand(ExecHelper.java:84)
	at com.rusefi.maintenance.DfuFlasher.executeDFU(DfuFlasher.java:231)
	at com.rusefi.maintenance.DfuFlasher.executeDfuAndPaintStatusPanel(DfuFlasher.java:208)
	at com.rusefi.maintenance.DfuFlasher.lambda$runDfuProgramming$5(DfuFlasher.java:186)
	at com.rusefi.maintenance.jobs.JobHelper.doJob(JobHelper.java:16)
	at com.rusefi.maintenance.DfuFlasher.lambda$runDfuProgramming$6(DfuFlasher.java:177)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: com.rusefi.maintenance.ErrorExecutingCommand
	at com.rusefi.maintenance.ExecHelper.executeCommand(ExecHelper.java:109)
	at com.rusefi.maintenance.ExecHelper.executeCommand(ExecHelper.java:80)
	... 6 more

```

--- a/java_console/ui/src/main/java/com/rusefi/DevicePane.java
+++ b/java_console/ui/src/main/java/com/rusefi/DevicePane.java
@@ -1,6 +1,7 @@
 package com.rusefi;
 
 import com.rusefi.core.preferences.storage.PersistentConfiguration;
+import com.rusefi.maintenance.DfuFlasher;
 import com.rusefi.maintenance.ProgramSelector;
 import com.rusefi.ui.UIContext;
 import com.rusefi.ui.basic.FirmwareRollbackController;
@@ -181,10 +182,12 @@ public class DevicePane {
 
     static String bootloaderGuidance(final SessionState state) {
         if (state == SessionState.DEVICE_IN_DFU) {
-            // DFU flashing is Windows-only in rusEFI (STM32_Programmer_CLI.exe); elsewhere it's a dead end.
-            return FileLog.isWindows()
-                ? "Board is in the DFU bootloader — click Update Firmware to flash."
-                : "Board is in the DFU bootloader. DFU flashing requires Windows — power-cycle to exit, or use OpenBLT.";
+            if (FileLog.isLinux()) {
+                return "Board is in the DFU bootloader - click Update Firmware to flash with dfu-util.";
+            }
+            return DfuFlasher.isDfuProgrammingSupported()
+                ? "Board is in the DFU bootloader - click Update Firmware to flash."
+                : "Board is in the DFU bootloader. DFU flashing is not supported on this platform.";
         }
         return "Board is in the OpenBLT bootloader — click Update Firmware to flash.";
     }

--- a/java_console/ui/src/main/java/com/rusefi/maintenance/DfuFlasher.java
+++ b/java_console/ui/src/main/java/com/rusefi/maintenance/DfuFlasher.java
@@ -21,8 +21,10 @@ import javax.swing.*;
 import java.awt.*;
 import java.io.*;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
 
@@ -32,6 +34,8 @@ import java.util.concurrent.atomic.AtomicReference;
 public class DfuFlasher {
     private static final String DFU_CMD_TOOL_LOCATION = Launcher.TOOLS_PATH + File.separator + "STM32_Programmer_CLI/bin";
     private static final String DFU_CMD_TOOL = "STM32_Programmer_CLI.exe";
+    private static final String DFU_UTIL = "dfu-util";
+    private static final String STM32_DFU_USB_ID = "0483:df11";
     private static final String WMIC_DFU_QUERY_COMMAND = "powershell -NoProfile -Command \"Get-CimInstance Win32_PnPEntity -Filter \\\"Caption like '%STM32%' and Caption like '%Bootloader%'\\\" | Select-Object Caption, ConfigManagerErrorCode | Format-List\"";
     private static final String WMIC_DFU_QUERY_H7_COMMAND = "powershell -NoProfile -Command \"Get-CimInstance Win32_PnPEntity -Filter \\\"Caption like '%DFU%' and Caption like '%FS Mode%'\\\" | Select-Object Caption, ConfigManagerErrorCode | Format-List\"";
     /**
@@ -42,6 +46,10 @@ public class DfuFlasher {
 
     public static boolean haveBootloaderBinFile() {
         return new File(FindFileHelper.findBootloaderFile()).exists();
+    }
+
+    public static boolean isDfuProgrammingSupported() {
+        return FileLog.isWindows() || FileLog.isLinux();
     }
 
     public static boolean doAutoDfu(
@@ -96,16 +104,20 @@ public class DfuFlasher {
             return false;
         }
 
+        if (!isDfuProgrammingSupported()) {
+            callbacks.logLine("DFU programming is not supported on " + FileLog.getOsName());
+            return false;
+        }
+
+        // Do not reboot a working ECU into DFU if the external Linux programmer is unavailable.
+        if (FileLog.isLinux() && !executeDfuUtilCommand(getDfuUtilVersionCommand(), callbacks)) {
+            return false;
+        }
+
         AtomicBoolean isSignatureValidated = rebootToDfu(parent, port, callbacks, Integration.CMD_REBOOT_DFU);
         if (isSignatureValidated == null)
             return false;
         if (isSignatureValidated.get()) {
-            if (!FileLog.isWindows()) {
-                callbacks.logLine("Switched to DFU mode!");
-                callbacks.logLine("rusEFI console can only program on Windows");
-                return false;
-            }
-
             timeForDfuSwitch(callbacks);
             final String fileName = firmwareBinFile != null
                 ? firmwareBinFile
@@ -188,6 +200,19 @@ public class DfuFlasher {
     }
 
     private static void runDfuErase(UpdateOperationCallbacks callbacks) {
+        if (FileLog.isLinux()) {
+            if (!executeDfuUtilCommand(getDfuUtilEraseCommand(), callbacks)) {
+                callbacks.error();
+            }
+            return;
+        }
+
+        if (!FileLog.isWindows()) {
+            callbacks.logLine("DFU erase is not supported on " + FileLog.getOsName());
+            callbacks.error();
+            return;
+        }
+
         try {
             ExecHelper.executeCommand(DFU_CMD_TOOL_LOCATION,
                     getDfuEraseCommand(),
@@ -241,7 +266,12 @@ public class DfuFlasher {
     }
 
     private static boolean executeDFU(UpdateOperationCallbacks callbacks, String firmwareBinFile,
-                                      ConnectedEcuTarget connectedEcuTarget) {
+                                       ConnectedEcuTarget connectedEcuTarget) {
+        if (!isDfuProgrammingSupported()) {
+            callbacks.logLine("DFU programming is not supported on " + FileLog.getOsName());
+            return false;
+        }
+
         // Refuse to silently flash firmware built for a different board (brick guard, [tag:better_ux_for_flashing]).
         if (!MaintenanceUtil.confirmFirmwareMatchesBoard(firmwareBinFile, callbacks, connectedEcuTarget)) {
             callbacks.logLine("Firmware update aborted — firmware/board mismatch.");
@@ -249,8 +279,21 @@ public class DfuFlasher {
         }
         boolean driverIsHappy = detectSTM32BootloaderDriverState(callbacks, connectedEcuTarget);
         if (!driverIsHappy) {
-            callbacks.logLine("*** DRIVER ERROR? *** Did you have a chance to try 'Install Drivers' button on top of rusEFI console start screen?");
+            if (FileLog.isLinux()) {
+                callbacks.logLine("STM32 DFU device 0483:df11 was not detected. Check the USB connection and lsusb access.");
+            } else {
+                callbacks.logLine("*** DRIVER ERROR? *** Did you have a chance to try 'Install Drivers' button on top of rusEFI console start screen?");
+            }
             return false;
+        }
+
+        if (FileLog.isLinux()) {
+            if (!executeDfuUtilCommand(getDfuUtilWriteCommand(firmwareBinFile), callbacks)) {
+                return false;
+            }
+
+            callbacks.logLine("SUCCESS!");
+            return true;
         }
 
         StringBuffer stdout = new StringBuffer();
@@ -289,7 +332,7 @@ public class DfuFlasher {
         if (!FileLog.isWindows()) {
             // The WMIC/driver check below is Windows-only. On Linux the STM32 system bootloader is
             // visible directly over USB as 0483:df11 ("STM Device in DFU Mode"), so probe lsusb so the
-            // console at least reflects the DFU state (flashing it is still Windows-only, see [tag:better_ux_for_flashing]).
+            // console reflects the DFU state before the Linux dfu-util flasher runs.
             return FileLog.isLinux() && detectStm32DfuViaLsusb(callbacks);
         }
         // #9714: a universal bundle's is_h7 property can't cover every board, so trust the connected
@@ -310,19 +353,32 @@ public class DfuFlasher {
     // via lsusb so a board reset into DFU surfaces in the console instead of leaving the watchdog spinning
     // on the vanished serial port. [tag:better_ux_for_flashing]
     private static boolean detectStm32DfuViaLsusb(UpdateOperationCallbacks callbacks) {
+        Process process = null;
         try {
-            Process process = new ProcessBuilder("lsusb").redirectErrorStream(true).start();
+            process = new ProcessBuilder("lsusb").redirectErrorStream(true).start();
+            if (!process.waitFor(3, TimeUnit.SECONDS)) {
+                process.destroyForcibly();
+                process.waitFor(2, TimeUnit.SECONDS);
+                callbacks.logLine("lsusb DFU probe timed out");
+                return false;
+            }
+
             StringBuilder output = new StringBuilder();
-            try (java.io.BufferedReader reader =
-                     new java.io.BufferedReader(new java.io.InputStreamReader(process.getInputStream()))) {
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()))) {
                 String line;
                 while ((line = reader.readLine()) != null) {
                     output.append(line).append('\n');
                 }
             }
-            process.waitFor(3, java.util.concurrent.TimeUnit.SECONDS);
             return output.toString().toLowerCase().contains("0483:df11");
-        } catch (Exception e) {
+        } catch (InterruptedException e) {
+            if (process != null) {
+                process.destroyForcibly();
+            }
+            Thread.currentThread().interrupt();
+            callbacks.logLine("lsusb DFU probe interrupted");
+            return false;
+        } catch (IOException e) {
             callbacks.logLine("lsusb DFU probe failed: " + e);
             return false;
         }
@@ -366,6 +422,94 @@ public class DfuFlasher {
 
     private static String getDfuEraseCommand() {
         return DFU_CMD_TOOL_LOCATION + "/" + DFU_CMD_TOOL + " -c port=usb1 -e all";
+    }
+
+    static List<String> getDfuUtilWriteCommand(String fileName) {
+        return Arrays.asList(
+            DFU_UTIL,
+            "--device", STM32_DFU_USB_ID,
+            "--alt", "0",
+            "--dfuse-address", "0x08000000:leave",
+            "--download", new File(fileName).getAbsolutePath()
+        );
+    }
+
+    static List<String> getDfuUtilEraseCommand() {
+        return Arrays.asList(
+            DFU_UTIL,
+            "--device", STM32_DFU_USB_ID,
+            "--alt", "0",
+            "--dfuse-address", "0x08000000:mass-erase:force"
+        );
+    }
+
+    private static List<String> getDfuUtilVersionCommand() {
+        return Arrays.asList(DFU_UTIL, "--version");
+    }
+
+    private static boolean executeDfuUtilCommand(List<String> command, UpdateOperationCallbacks callbacks) {
+        callbacks.logLine("Executing command=" + String.join(" ", command));
+
+        Process process = null;
+        Thread outputThread = null;
+        StringBuffer output = new StringBuffer();
+        try {
+            process = new ProcessBuilder(command).redirectErrorStream(true).start();
+            final Process runningProcess = process;
+            outputThread = new Thread(() -> {
+                try (BufferedReader reader = new BufferedReader(new InputStreamReader(runningProcess.getInputStream()))) {
+                    String line;
+                    while ((line = reader.readLine()) != null) {
+                        callbacks.logLine(line);
+                        output.append(line).append('\n');
+                    }
+                } catch (IOException e) {
+                    callbacks.logLine("dfu-util output error: " + e);
+                }
+            }, DfuFlasher.class.getSimpleName() + " dfu-util output");
+            outputThread.setDaemon(true);
+            outputThread.start();
+
+            if (!process.waitFor(3, TimeUnit.MINUTES)) {
+                process.destroyForcibly();
+                process.waitFor(2, TimeUnit.SECONDS);
+                outputThread.join(2000);
+                callbacks.logLine("dfu-util timed out after 3 minutes");
+                return false;
+            }
+            outputThread.join(2000);
+
+            if (process.exitValue() == 0) {
+                return true;
+            }
+
+            callbacks.logLine("dfu-util exited with code " + process.exitValue());
+            String lowerOutput = output.toString().toLowerCase();
+            if (lowerOutput.contains("libusb_error_access")
+                || lowerOutput.contains("permission denied")
+                || lowerOutput.contains("cannot open dfu device")) {
+                callbacks.logLine("Grant this user USB access to 0483:df11 with an appropriate udev rule, then reconnect the ECU.");
+            }
+            return false;
+        } catch (IOException e) {
+            callbacks.logLine("Unable to run dfu-util. Install dfu-util and make sure it is available on PATH: " + e);
+            return false;
+        } catch (InterruptedException e) {
+            if (process != null) {
+                process.destroyForcibly();
+                try {
+                    process.waitFor(2, TimeUnit.SECONDS);
+                    if (outputThread != null) {
+                        outputThread.join(2000);
+                    }
+                } catch (InterruptedException ignored) {
+                    // Restore interruption below after best-effort process cleanup.
+                }
+            }
+            Thread.currentThread().interrupt();
+            callbacks.logLine("dfu-util was interrupted");
+            return false;
+        }
     }
 
     @NotNull

--- a/java_console/ui/src/main/java/com/rusefi/maintenance/ProgramSelector.java
+++ b/java_console/ui/src/main/java/com/rusefi/maintenance/ProgramSelector.java
@@ -73,12 +73,13 @@ public class ProgramSelector {
             executeJob(splitButton, mainButtonModeFor(targetPort), targetPort);
         });
 
-        // Keep the main button label in sync with whatever port is currently selected. The combo is
+        // Keep the main button label and enabled state in sync with whatever port is currently selected. The combo is
         // shared with the Connect tab and is repopulated/re-selected after apply() runs, so drive the
-        // text off selection changes (same source the action reads) rather than off apply()'s snapshot.
+        // control off selection changes (same source the action reads) rather than off apply()'s snapshot.
         comboPorts.addItemListener(e -> {
             if (e.getStateChange() == ItemEvent.SELECTED) {
                 refreshMainButtonText();
+                apply(connectivityContext.getCurrentHardware());
             }
         });
         refreshMainButtonText();
@@ -560,30 +561,34 @@ public class ProgramSelector {
         updateModeAndButton.setVisible(shouldShowFirmwareControls(
             currentHardware.isEmpty(), isJobRunning, additionalControlVisible));
 
-        boolean hasSerialPorts = !currentHardware.getKnownPorts().isEmpty();
+        final List<PortResult> knownPorts = currentHardware.getKnownPorts();
+        boolean hasSerialPorts = hasRealSerialPort(knownPorts);
         boolean hasDfuDevice = currentHardware.isDfuFound();
+        boolean supportsDfu = DfuFlasher.isDfuProgrammingSupported();
 
         JPopupMenu popupMenu = new JPopupMenu();
 
-        if (FileLog.isWindows()) {
-            boolean requireBlt = FindFileHelper.isObfuscated()
-                || isForeignBoardOnUniversalBundle(connectivityContext.getConnectedEcuTarget());
+        boolean requireBlt = FindFileHelper.isObfuscated()
+            || isForeignBoardOnUniversalBundle(connectivityContext.getConnectedEcuTarget());
+        boolean canUseDfu = supportsDfu && !requireBlt;
 
-            if (!requireBlt) {
-                if (hasSerialPorts) {
-                    addMenuItem(popupMenu, DFU_AUTO);
-                }
-                if (hasDfuDevice) {
-                    addMenuItem(popupMenu, DFU_MANUAL);
-                    addMenuItem(popupMenu, DFU_ERASE);
-                    if (DfuFlasher.haveBootloaderBinFile()) {
-                        addMenuItem(popupMenu, INSTALL_OPENBLT);
-                    }
-                }
+        if (canUseDfu) {
+            if (hasSerialPorts) {
+                addMenuItem(popupMenu, DFU_AUTO);
                 addMenuItem(popupMenu, DFU_SWITCH);
-                if (currentHardware.isStLinkConnected()) {
-                    addMenuItem(popupMenu, ST_LINK);
+            }
+            if (hasDfuDevice) {
+                addMenuItem(popupMenu, DFU_MANUAL);
+                addMenuItem(popupMenu, DFU_ERASE);
+                if (DfuFlasher.haveBootloaderBinFile()) {
+                    addMenuItem(popupMenu, INSTALL_OPENBLT);
                 }
+            }
+        }
+
+        if (FileLog.isWindows()) {
+            if (!requireBlt && currentHardware.isStLinkConnected()) {
+                addMenuItem(popupMenu, ST_LINK);
             }
             if (currentHardware.isPCANConnected()) {
                 addMenuItem(popupMenu, OPENBLT_CAN);
@@ -598,7 +603,8 @@ public class ProgramSelector {
         int menuItemCount = popupMenu.getComponentCount();
 
         splitButton.setPopupMenu(menuItemCount > 0 ? popupMenu : null);
-        splitButton.setMainButtonEnabled(hasSerialPorts && !isJobRunning);
+        splitButton.setMainButtonEnabled(shouldEnableMainButton(
+            hasSerialPorts, hasDfuDevice, isJobRunning, mainButtonModeFor(resolveFlashPort()), canUseDfu));
         splitButton.setArrowButtonEnabled(menuItemCount > 0 && !isJobRunning);
 
         // Keep the main-button mode/label in sync with the connection state too (not just combo changes):
@@ -607,6 +613,22 @@ public class ProgramSelector {
 
         AutoupdateUtil.trueLayoutAndRepaint(splitButton);
         AutoupdateUtil.trueLayoutAndRepaint(content);
+    }
+
+    static boolean hasRealSerialPort(List<PortResult> ports) {
+        return ports.stream().anyMatch(port -> port.type != SerialPortType.Dfu);
+    }
+
+    static boolean shouldEnableMainButton(
+        boolean hasSerialPorts,
+        boolean hasDfuDevice,
+        boolean jobRunning,
+        UpdateMode mode,
+        boolean supportsDfu
+    ) {
+        boolean targetAvailable = mode == DFU_MANUAL ? hasDfuDevice : hasSerialPorts;
+        boolean modeSupported = mode != DFU_MANUAL || supportsDfu;
+        return targetAvailable && modeSupported && !jobRunning;
     }
 
     static boolean shouldShowFirmwareControls(

--- a/java_console/ui/src/test/java/com/rusefi/DevicePaneTest.java
+++ b/java_console/ui/src/test/java/com/rusefi/DevicePaneTest.java
@@ -27,12 +27,13 @@ public class DevicePaneTest {
     public void dfuGuidanceMentionsDfuAndReflectsPlatformSupport() {
         String guidance = DevicePane.bootloaderGuidance(SessionState.DEVICE_IN_DFU);
         assertTrue(guidance.contains("DFU"), guidance);
-        if (FileLog.isWindows()) {
-            // DFU flashing works here — send the user to the update button
+        if (FileLog.isLinux()) {
+            assertTrue(guidance.contains("dfu-util"), guidance);
+            assertTrue(guidance.contains("Update Firmware"), guidance);
+        } else if (FileLog.isWindows()) {
             assertTrue(guidance.contains("Update Firmware"), guidance);
         } else {
-            // DFU flashing is Windows-only (STM32_Programmer_CLI.exe) — say so instead of a dead end
-            assertTrue(guidance.contains("Windows"), guidance);
+            assertTrue(guidance.contains("not supported"), guidance);
         }
     }
 

--- a/java_console/ui/src/test/java/com/rusefi/maintenance/DfuFlasherTest.java
+++ b/java_console/ui/src/test/java/com/rusefi/maintenance/DfuFlasherTest.java
@@ -1,0 +1,33 @@
+package com.rusefi.maintenance;
+
+import org.junit.jupiter.api.Test;
+
+import java.io.File;
+import java.util.Arrays;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DfuFlasherTest {
+    @Test
+    public void linuxWriteCommandTargetsStm32FlashAndLeavesDfu() {
+        File firmware = new File("directory with spaces", "rusefi.bin");
+
+        assertEquals(Arrays.asList(
+            "dfu-util",
+            "--device", "0483:df11",
+            "--alt", "0",
+            "--dfuse-address", "0x08000000:leave",
+            "--download", firmware.getAbsolutePath()
+        ), DfuFlasher.getDfuUtilWriteCommand(firmware.getPath()));
+    }
+
+    @Test
+    public void linuxEraseCommandRequiresExplicitMassEraseForce() {
+        assertEquals(Arrays.asList(
+            "dfu-util",
+            "--device", "0483:df11",
+            "--alt", "0",
+            "--dfuse-address", "0x08000000:mass-erase:force"
+        ), DfuFlasher.getDfuUtilEraseCommand());
+    }
+}

--- a/java_console/ui/src/test/java/com/rusefi/maintenance/ProgramSelectorTest.java
+++ b/java_console/ui/src/test/java/com/rusefi/maintenance/ProgramSelectorTest.java
@@ -11,6 +11,7 @@ import static com.rusefi.maintenance.ProgramSelector.mainButtonModeFor;
 import static com.rusefi.maintenance.ProgramSelector.resolveFlashPort;
 import static com.rusefi.maintenance.UpdateMode.*;
 import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
 import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertTrue;
@@ -144,5 +145,19 @@ public class ProgramSelectorTest {
     public void firmwareControlsStayVisibleWhileJobRunsWithoutDetectedHardware() {
         assertTrue(ProgramSelector.shouldShowFirmwareControls(true, true, false));
         assertTrue(ProgramSelector.shouldShowFirmwareControls(true, false, true));
+    }
+
+    @Test
+    public void syntheticDfuTargetIsNotTreatedAsSerialPort() {
+        assertFalse(ProgramSelector.hasRealSerialPort(Collections.singletonList(port(SerialPortType.Dfu))));
+        assertTrue(ProgramSelector.hasRealSerialPort(Collections.singletonList(port(SerialPortType.Ecu))));
+    }
+
+    @Test
+    public void mainDfuActionRequiresPlatformSupportButOpenBltDoesNot() {
+        assertTrue(ProgramSelector.shouldEnableMainButton(false, true, false, DFU_MANUAL, true));
+        assertFalse(ProgramSelector.shouldEnableMainButton(false, true, false, DFU_MANUAL, false));
+        assertTrue(ProgramSelector.shouldEnableMainButton(true, false, false, OPENBLT_MANUAL, false));
+        assertFalse(ProgramSelector.shouldEnableMainButton(true, false, true, OPENBLT_MANUAL, true));
     }
 }

